CRM/Mailing/BAO/Mailing.php

```diff
@@ -109,46 +109,227 @@ public static function getRecipientsCount($job_id, $mailing_id = NULL) {
-    $mode = NULL) {
+  public static function getMailRecipients($mailingObj) {
+    $mailing = CRM_Mailing_BAO_Mailing::getTableName();
+    $mailingID = $mailingObj->id;
+
     $mailingGroup = new CRM_Mailing_DAO_MailingGroup();
+    $recipientsGroup = $excludeSmartGroupIDs = $includeSmartGroupIDs = array();
+    $sql = " SELECT GROUP_CONCAT(entity_id SEPARATOR \",\") as group_ids, group_type
+      FROM civicrm_mailing_group
+      WHERE mailing_id = %1 AND entity_table = 'civicrm_group'
+      GROUP BY group_type
+    ";
+    $dao = CRM_Core_DAO::executeQuery($sql, array(1 => array($mailingID, 'Int')));
+    while ($dao->fetch()) {
+      $recipientsGroup[$dao->group_type] = explode(',', $dao->group_ids);
+    }
 
-    $mailing = CRM_Mailing_BAO_Mailing::getTableName();
-    $job = CRM_Mailing_BAO_MailingJob::getTableName();
-    $mg = CRM_Mailing_DAO_MailingGroup::getTableName();
-    $eq = CRM_Mailing_Event_DAO_Queue::getTableName();
+    // there is no need to proceed further as no mailing group is selected to include recipients,
+    // but before return clear the mailing recipients populated earlier since as per current params no group is selected
+    if (empty($recipientsGroup['Include'])) {
+      $sql = " DELETE FROM civicrm_mailing_recipients WHERE  mailing_id = %1 ";
+      CRM_Core_DAO::executeQuery($sql, array(1 => array($mailingID, 'Integer')));
+      return;
+    }
+
+    list($location_filter, $order_by) = self::getLocationFilterAndOrderBy($mailingObj->email_selection_method, $mailingObj->location_type_id);
+
+    // get all the saved searches AND hierarchical groups
+    // and load them in the cache
+    foreach ($recipientsGroup as $groupType => $groupIDs) {
+      $sql = sprintf("
+       SELECT *
+       FROM  civicrm_group g
+       WHERE id IN (%s) AND ( saved_search_id != 0 OR saved_search_id IS NOT NULL OR children IS NOT NULL )
+      ", implode(',', $groupIDs));
+
+      $groupDAO = CRM_Core_DAO::executeQuery($sql);
+      while ($groupDAO->fetch()) {
+        if ($groupDAO->cache_date == NULL) {
+          CRM_Contact_BAO_GroupContactCache::load($groupDAO);
+        }
+        if ($groupType == 'Include') {
+          $includeSmartGroupIDs[] = $groupDAO->id;
+        }
+        else {
+          $excludeSmartGroupIDs[] = $groupDAO->id;
+        }
+      }
+    }
+
+    // Create a temp table for contact exclusion.
+    $excludeTempTablename = "excluded_recipients_mailing_temp";
+    $includedTempTablename = "included_recipients_mailing_temp";
+    $mailingGroup->query(
+      "CREATE TEMPORARY TABLE $excludeTempTablename
+            (contact_id int primary key)
+            ENGINE=HEAP"
+    );
+    if (!empty($recipientsGroup['Exclude'])) {
+      $sql = sprintf("
+        INSERT INTO %s (contact_id)
+          SELECT DISTINCT gc.contact_id
+          FROM  civicrm_group_contact gc
+          WHERE gc.status = 'Added' AND gc.id IN (%s)", $excludeTempTablename, implode(',', $recipientsGroup['Exclude']));
```

Comment:
Are the names of Groups sanitized?

Reply:
FWIW, the query-builder can handle `INSERT...SELECT...` with validation/escaping/interpolation:

```php
CRM_Utils_SQL_Select::from('civicrm_group_contact')
  ->select('DISTINCT gc.contact_id')
  ->where('gc.status = "Added" AND gc.id IN (#groups)')
  ->param('#groups', $recipientsGroup['Exclude'])
  ->insertInto($excludeTempTablename, array('contact_id'))
  ->execute();
```